src/main/java/org/jenkinsci/plugins/workflow/support/steps/build/BuildUpstreamNodeAction.java

```diff
@@ -0,0 +1,53 @@
+package org.jenkinsci.plugins.workflow.support.steps.build;
+
+import hudson.model.InvisibleAction;
+import hudson.model.Run;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+
+import java.util.Objects;
+
+/**
+ * Attached to newly-created builds in order to point back to the triggering FlowNode.
+ *
+ * We annotate the downstream build instead of upstream in order to support no-wait builds,
+ * because the downstream run id is not available when they're still queued.
+ *
+ * Needed for Blue Ocean to annotate the correct step.
+ * @see: https://issues.jenkins-ci.org/browse/JENKINS-38339
+ */
+public class BuildUpstreamNodeAction extends InvisibleAction {
+
+    private final String upstreamNodeId;
+    private final String upstreamRunId;
+
+    public BuildUpstreamNodeAction(FlowNode node, Run<?, ?> invokingRun) {
+        this.upstreamNodeId = node.getId();
+        this.upstreamRunId = invokingRun.getExternalizableId();
+    }
+
+    public String getUpstreamNodeId() {
+        return upstreamNodeId;
+    }
+
+    public String getUpstreamRunId() {
+        return upstreamRunId;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
Probably unnecessary to override `equals` & `hashCode`.

Reply:
I like to be on the safe side when I know something is going in and out of collections, and there are places in the API docs that specifically mention making sure you have `equals` and `hashCode` on `Action`s